File: XBMCSkinEditor/Controls/SkinImages.vb
Imports System.Drawing.Imaging

Public Class ImageFrame

    Private _Frame As Drawing.Image
    Private _Delay As Integer

    Public Property Frame() As Image
        Get
            Return _Frame
        End Get
        Set(ByVal Value As Image)
            _Frame = Value
        End Set
    End Property

    Public Property Delay() As Integer
        Get
            Return _Delay
        End Get
        Set(ByVal Value As Integer)
            _Delay = Value
        End Set
    End Property

End Class

Public Class ImageFrames
    Inherits System.Collections.CollectionBase

    Public Overridable Function Add(ByVal Frame As ImageFrame) As Integer
        MyBase.List.Add(Frame)
    End Function

    Default Public Overridable Property Item(ByVal Index As Integer) As ImageFrame
        Get
            Return DirectCast(MyBase.List.Item(Index), ImageFrame)
        End Get
        Set(ByVal Frame As ImageFrame)
            MyBase.List.Item(Index) = Frame
        End Set
    End Property

End Class

Public Class SkinImage

    Public Event ShowFrame(ByVal Image As Image)

    Private _Loops As Integer
    Private _Frames As New ImageFrames

    Private _Timer As Timer
    Private _Frame As Integer = 0
    Private _Loop As Integer = 1

    Public Property Loops() As Integer
        Get
            Return _Loops
        End Get
        Set(ByVal Value As Integer)
            _Loops = Value
        End Set
    End Property

    Public Property Frames() As ImageFrames
        Get
            Return _Frames
        End Get
        Set(ByVal Value As ImageFrames)
            _Frames = Value
        End Set
    End Property

    Public Sub Load(ByVal FileName As String)

        Dim bytDelays() As Byte = Nothing
        Dim objImage As Image = Image.FromFile(FileName)
        Dim objDimension As New FrameDimension(objImage.FrameDimensionsList(0))

        _Frame = 0
        _Loops = 0

        Dim objPropertyItem As PropertyItem
        For Each objPropertyItem In objImage.PropertyItems
            Select Case objPropertyItem.Id
                Case 20736
                    bytDelays = objImage.GetPropertyItem(objPropertyItem.Id).Value
                Case 20737
                    _Loops = BitConverter.ToInt16(objImage.GetPropertyItem(objPropertyItem.Id).Value, 0)
            End Select
        Next objPropertyItem

        _Frames.Clear()
        For intLoop As Integer = 0 To objImage.GetFrameCount(objDimension) - 1

            Dim objImageFrame As New ImageFrame
            Dim objMemoryStream As New System.IO.MemoryStream()

            objImage.SelectActiveFrame(objDimension, intLoop)
            objImage.Save(objMemoryStream, System.Drawing.Imaging.ImageFormat.Png)

            If bytDelays Is Nothing Then
                objImageFrame.Delay = 0
            Else
                objImageFrame.Delay = BitConverter.ToInt32(bytDelays, intLoop * 4) * 10
            End If

            objImageFrame.Frame = System.Drawing.Image.FromStream(objMemoryStream)
            _Frames.Add(objImageFrame)

        Next intLoop

    End Sub

    Public Sub Animate()

        If _Frames.Count > 0 Then

            _Loop = _Loops
            _Frame = 0
            _Timer = New Timer
            AddHandler _Timer.Tick, AddressOf TimerTick

            _Timer.Interval = 100 '_Frames.Item(0).Delay
            _Timer.Enabled = True

        End If

    End Sub

    Private Sub TimerTick(ByVal sender As Object, ByVal e As System.EventArgs)
        RaiseEvent ShowFrame(_Frames.Item(_Frame).Frame)
        _Frame += 1
        If _Frame = _Frames.Count Then _Frame = 0
    End Sub

End Class

File: branches/XBMCSkinEditor_2008/XBMCSkinEditor/Classes/SkinImages.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;

namespace XBMCSkinEditor.Classes
{
    using System.Drawing.Imaging;

    public class ImageFrame
    {

        private Image frame;
        private int delay;

        public Image Frame
        {
            get { return frame; }
            set { frame = value; }
        }

        public int Delay
        {
            get { return delay; }
            set { delay = value; }
        }

    }

    public class SkinImage
    {

        public event ShowFrameEventHandler ShowFrame;
        public delegate void ShowFrameEventHandler(Image Image);

        private int _Loops;
        private System.Collections.ArrayList frames = new System.Collections.ArrayList();
        
        //private System.Timers.Timer _Timer;
        private int frame = 0;
        private int loops = 1;

        public int Loops
        {
            get { return this.loops; }
            set { this.loops = value; }
        }

        public System.Collections.ArrayList Frames
        {

            get { return this.frames; }
            set { this.frames = value; }
        }

        public void Load(string FileName)
        {
            
        byte[] bytDelays = null;
        Image objImage = Image.FromFile(FileName);
        //FrameDimension objDimension = new FrameDimension(objImage.FrameDimensionsList(0));
        
        this.frame = 0;
        _Loops = 0;
        
        foreach ( PropertyItem objPropertyItem in objImage.PropertyItems) {
            switch (objPropertyItem.Id) {
                case 20736:
                    bytDelays = objImage.GetPropertyItem(objPropertyItem.Id).Value;
                    break;
                case 20737:
                    _Loops = BitConverter.ToInt16(objImage.GetPropertyItem(objPropertyItem.Id).Value, 0);
                    break;
            }
        }
        
        this.frames.Clear();
        /*for (int intLoop = 0; intLoop <= objImage.GetFrameCount(objDimension) - 1; intLoop++) {
            
            ImageFrame objImageFrame = new ImageFrame();
            System.IO.MemoryStream objMemoryStream = new System.IO.MemoryStream();
            
            objImage.SelectActiveFrame(objDimension, intLoop);
            objImage.Save(objMemoryStream, System.Drawing.Imaging.ImageFormat.Png);
            
            if (bytDelays == null) {
                objImageFrame.Delay = 0;
            }
            else {
                objImageFrame.Delay = BitConverter.ToInt32(bytDelays, intLoop * 4) * 10;
            }
            
            objImageFrame.Frame = System.Drawing.Image.FromStream(objMemoryStream);
            _Frames.Add(objImageFrame);
            
        }*/
        
    }
        /*
        public void Animate()
        {

            if (_Frames.Count > 0)
            {

                _Loop = _Loops;
                _Frame = 0;
                _Timer = new Timer();
                _Timer.Tick += TimerTick;

                _Timer.Interval = 100;
                //_Frames.Item(0).Delay
                _Timer.Enabled = true;

            }

        }
        */
        private void TimerTick(object sender, System.EventArgs e)
        {
            if (ShowFrame != null)
            {
                //ShowFrame(this.frames[this.frame].Frame);
            }

            this.frame += 1;
            if (this.frame == this.frames.Count)
                this.frame = 0;
        }
    }


}

